wasm/src/lib.rs

```diff
@@ -0,0 +1,16 @@
+use css_inline as rust_inline;
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+pub fn inline(
+    html: &str,
+    remove_style_tags: Option<bool>,
+) -> Result<String, JsValue> {
+    let options = rust_inline::InlineOptions {
+        remove_style_tags: remove_style_tags.unwrap_or(false),
+        base_url: None,
```

Comment:
It would be nice to provide the ability to pass all config values to `InlineOptions`, not sure what is the canonical way to do so in JS - is it an extra "config" object like it is done in `juice`? Or is it better to have multiple arguments?

Reply:
will do  
a config object would be the preferred way to go as that's the only way to provide named arguments :)

Reply:
although how would something like that be written? :p

I tried

```rust
#[wasm_bindgen]
pub fn inline(
    html: &str,
    options: rust_inline::InlineOptions,
) -> Result<String, JsValue> {
```

but understandably got 
```
`the trait `wasm_bindgen::convert::traits::FromWasmAbi` is not implemented for `css_inline::InlineOptions<'_>`
```

Reply:
Cool! I'll post an update regarding the struct conversion soon :) Need to do some more experiments, but the general idea could be taken from [here](https://rustwasm.github.io/wasm-bindgen/reference/arbitrary-data-with-serde.html) + another newtype wrapper to avoid `unwrap`

Reply:
To use an object for "options" argument we need to:

1. Use `JsValue` as the argument type

```rust
#[wasm_bindgen]
pub fn inline(html: &str, options: JsValue) -> Result<String, JsValue> {
```

2. Convert this `JsValue` to `rust_inline::InlineOptions<'_>`
I think that the easiest way is to add an intermediate `Options` struct, and [parse](https://rustwasm.github.io/wasm-bindgen/reference/arbitrary-data-with-serde.html#receive-it-from-javascript-with-jsvalueinto_serde) `JsValue` into it with `serde`:

```rust
#[macro_use]
extern crate serde_derive;

#[derive(Deserialize)]
#[serde(default)]
pub struct Options {
    inline_style_tags: bool,
    remove_style_tags: bool,
    base_url: Option<String>,
    load_remote_stylesheets: bool,
    extra_css: Option<String>,
}

impl Default for Options {
    fn default() -> Self {
        Options {
            inline_style_tags: true,
            remove_style_tags: false,
            base_url: None,
            load_remote_stylesheets: true,
            extra_css: None,
        }
    }
}
```

Then to avoid `unwrap` add another newtype wrapper and `From` implementation for `JsValue`:

```rust
struct SerdeError(serde_json::Error);

impl From<SerdeError> for JsValue {
    fn from(error: SerdeError) -> Self {
        JsValue::from_str(error.0.to_string().as_str())
    }
}
```

So we can parse simple JSON values like this:

```rust
#[wasm_bindgen]
pub fn inline(html: &str, options: JsValue) -> Result<String, JsValue> {
    let options: Options = options.into_serde().map_err(SerdeError)?;
    // ...
}
```

3. Convert this intermediate struct to `InlineOptions`
The conversion is fallible, so we need to implement `TryFrom`:

```rust
impl TryFrom<Options> for rust_inline::InlineOptions<'_> {
    type Error = JsValue;

    fn try_from(value: Options) -> Result<Self, Self::Error> {
        Ok(rust_inline::InlineOptions {
            inline_style_tags: value.inline_style_tags,
            remove_style_tags: value.remove_style_tags,
            base_url: parse_url(value.base_url)?,
            load_remote_stylesheets: value.load_remote_stylesheets,
            extra_css: value.extra_css.map(Cow::Owned),
        })
    }
}
```

(I used `parse_url` and `UrlError` + `From` implementation you added)
And the final version might look like this:

```rust
#[wasm_bindgen]
pub fn inline(html: &str, options: JsValue) -> Result<String, JsValue> {
    let options: Options = options.into_serde().map_err(SerdeError)?;
    let inliner = rust_inline::CSSInliner::new(options.try_into()?);
    Ok(inliner.inline(html).unwrap())
}
```

The downside is having some code duplication between the original crate options and `Options`, but I think it is acceptable for now since there is an extra logical step involved (those values are 1:1 mapped from JS and then URL parsing / `Cow` wrapping is applied). So, I think we can start like this and then maybe add some refactorings :)

Reply:
Ah, one extra thing - `options` probably should be `Option<JsValue>`, but according to the [docs](https://rustwasm.github.io/wasm-bindgen/reference/types/jsvalue.html#jsvalue) it is not possible at the moment, but I think that there could be a workaround like this:

```rust
#[wasm_bindgen]
pub fn inline(html: &str, options: JsValue) -> Result<String, JsValue> {
    let options: Options = if !options.is_undefined() {
        options.into_serde().map_err(SerdeError)?
    } else {
        Options::default()
    };
    let inliner = rust_inline::CSSInliner::new(options.try_into()?);
    Ok(inliner.inline(html).unwrap())
}
```

I checked it locally - works with and without second argument

Reply:
Update: I copy-pasted my old code and there was `unwrap` - I updated it with `map_err` as it was in my earlier comments

Reply:
that works! tested locally as well :) 

there seems to be a bit of tradeoff here that the typescript bindings that end up being generated this way end up with `options` as `any`; 

```ts
/ **
* @param {string} html
* @param {any} options
* @returns {string}
*/
export function inline(html: string, options: any): string;
```

it looks like a limitation that's being discussed/explored? https://github.com/rustwasm/wasm-bindgen/issues/1591

Might need to go into the definition file to manually fiddle with the generated typescript definitions before publishing until wasm-bindgen figures this out

<details>
I tried the suggestion from https://github.com/rustwasm/wasm-bindgen/issues/1591#issuecomment-501784387 with

```rust
#[wasm_bindgen]
#[derive(Deserialize)]
#[serde(default)]
pub struct Options {
    pub inline_style_tags: bool,
    pub remove_style_tags: bool,
    pub base_url: Option<String>,
    pub load_remote_stylesheets: bool,
    pub extra_css: Option<String>,
}
```
but that fails with "the trait `std::marker::Copy` is not implemented for `std::string::String`"


Removing `pub` on the strings makes it sort of work

```rust
#[wasm_bindgen]
#[derive(Deserialize)]
#[serde(default)]
pub struct Options {
    pub inline_style_tags: bool,
    pub remove_style_tags: bool,
    base_url: Option<String>,
    pub load_remote_stylesheets: bool,
    extra_css: Option<String>,
}
```

this generates

```ts
export class Options {
  free(): void;
/**
* @returns {boolean}
*/
  inline_style_tags: boolean;
/**
* @returns {boolean}
*/
  load_remote_stylesheets: boolean;
/**
* @returns {boolean}
*/
  remove_style_tags: boolean;
}
```

![image](https://user-images.githubusercontent.com/743976/87257689-01dc9700-c46b-11ea-9ed7-a18116ac186c.png)

it is still weird that a class is generated rather than an interface/dictionary; and weird that there's a `free` method

</details>

Reply:
Hm, I need to do some research on this matter :) Would it work if we'll write proper TS types manually?

Reply:
It should, although I'm not sure how to tell rust where to look for the manually written TS types?

Reply:
Probably we can use some sections from [here](https://rustwasm.github.io/docs/wasm-bindgen/reference/attributes/on-rust-exports/index.html). For example [skip](https://rustwasm.github.io/docs/wasm-bindgen/reference/attributes/on-rust-exports/skip_typescript.html) TS generation for `inline`:

```rust
#[wasm_bindgen(skip_typescript)]
pub fn inline(html: &str, options: &JsValue) -> Result<String, JsValue> {
    // ...
}
```

And define it [manually](https://rustwasm.github.io/docs/wasm-bindgen/reference/attributes/on-rust-exports/typescript_custom_section.html):

```rust
#[wasm_bindgen(typescript_custom_section)]
const INLINE: &'static str = r#"
... some TS definitions
"#;
```

> although I'm not sure how to tell rust where to look for the manually written TS types?

Hope that it sheds some light on this if not - please, let me know if I missed something :)

Reply:
cool!